gobblin-config-management/src/main/java/gobblin/config/client/ConfigClient.java

```diff
@@ -0,0 +1,257 @@
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+import com.typesafe.config.Config;
+
+
+/**
+ * ConfigClient is the consumer facing class for retrieving configuration against all configuration stores.
+ * @author mitu
+ *
+ */
+public class ConfigClient {
+
+  public static enum VERSION_STABILITY_POLICY {
+    RAISE_ERROR,
+    CACHE_CONFIG_IN_MEMORY
+  }
+
+  private final VERSION_STABILITY_POLICY policy;
+
+  // key is the store ROOT, must use TreeMap
+  private final TreeMap<URI, ConfigStoreAccessor> configStoreMap = new TreeMap<URI, ConfigStoreAccessor>(
+      new URIComparator());
+
+  private ConfigClient(VERSION_STABILITY_POLICY policy) {
+    this.policy = policy;
+  }
+
+  /**
+   * @param policy - the policy control the behavior when this configuration client try to connect to configuration store.
+   * if policy is RAISE_ERROR and this client try to connect to any {@gobblin.config.configstore.ConfigStore} which is not
+   * {@gobblin.config.configstore.ConfigStoreWithStableVersion}, the RuntimeException will be thrown
+   * 
+   * if policy is CACHE_CONFIG_IN_MEMORY and this client try to connect to any {@gobblin.config.configstore.ConfigStore} which is not
+   * {@gobblin.config.configstore.ConfigStoreWithStableVersion}, the consumer need to cache the retrieved configuration in memory,
+   * otherwise, the same query for the same URI against same configuration store may result different result.
+   * @return - the Configuration Client object
+   */
+  public static ConfigClient createConfigClientWithStableVersion(VERSION_STABILITY_POLICY policy) {
+    return new ConfigClient(policy);
+  }
+
+  /**
+   * 
+   * @return the {@ConfigClient} with policy set to VERSION_STABILITY_POLICY.RAISE_ERROR
+   */
+  public static ConfigClient createDefaultConfigClient() {
+    // create with stable versions
+    return createConfigClientWithStableVersion(VERSION_STABILITY_POLICY.RAISE_ERROR);
+  }
+
+  private ConfigStoreAccessor getConfigStoreAccessor(URI uri) throws Exception {
+    if (this.configStoreMap.containsKey(uri)) {
+      return this.configStoreMap.get(uri);
+    }
+
+    ConfigStoreFactory<ConfigStore> csFactory = this.getConfigStoreFactory(uri);
+    ConfigStore cs = csFactory.createConfigStore(uri);
+    
+    if(!(cs instanceof ConfigStoreWithStableVersion)){
+      if(this.policy == VERSION_STABILITY_POLICY.RAISE_ERROR){
+        throw new Exception(String.format("Try to connect to unstable config store ", cs.getStoreURI()));
+      }
+    }
+    
+    // ConfigStoreFactory scheme name could be different than configStore's StoreURI's scheme name
+    URI csRoot = cs.getStoreURI();
+    URI key = new URI(csFactory.getScheme(), csRoot.getAuthority(), csRoot.getPath(), csRoot.getQuery(), csRoot.getFragment());
```

Comment:
We could just throw an error if the scheme is different. Is there a point to allowing this use case?